### ISSUE TITLE
Disposable of performing action in MainScheduler does not handled

### ISSUE THREAD

**Issue by valeriashershen** [ReactiveX/RxSwift#2140]
**Short description of the issue**:

Disposable of performing action in `MainScheduler` does not handled. So there is no possibility to cancel already started work in `MainScheduler`, however in `ConcurrentMainScheduler` such functionality is available.

**Self contained code example that reproduces the issue**:

```swift
import RxSwift

final class MyDisposable: Disposable {

    func dispose() {
        print("Disposed: MyDisposable")
    }

    deinit { dispose() }
}

// this instance is available in global scope
let disposeBag = DisposeBag() 

let myWorkBlock: () -> Disposable = {
    // ... some work
    print("Action performing")
    return MyDisposable()
}

DispatchQueue.global(qos: .background).async {
    MainScheduler.instance.schedule(Void()) { (_) -> Disposable in
        myWorkBlock()
    }.disposed(by: disposeBag)
}
```

**Expected outcome**:
Expected console output: 
```
Action performing
```
Disposable returned by scheduling action in `MainScheduler` is retained.

**What actually happens**:
Actual console output: 
```
Action performing
Disposed: MyDisposable
```
Disposable returned by scheduling action in `MainScheduler` does not retained. But if we use `ConcurrentMainScheduler` the disposable remains alive. 
After `RxSwift` source code investigation a difference in implementation of `MainScheduler` and `ConcurrentMainScheduler`  was found.
Specifically, sheduling method is implemented in different ways:
In `MainScheduler`:
```swift
override func scheduleInternal<StateType>(_ state: StateType, action: @escaping (StateType) -> Disposable) -> Disposable {
    let previousNumberEnqueued = increment(self.numberEnqueued)

    if DispatchQueue.isMain && previousNumberEnqueued == 0 {
        let disposable = action(state)
        decrement(self.numberEnqueued)
        return disposable
    }

    let cancel = SingleAssignmentDisposable()

    self._mainQueue.async {
        if !cancel.isDisposed {
            _ = action(state)
        }

        decrement(self.numberEnqueued)
    }

    return cancel
}
```
In `ConcurrentMainScheduler`:
```swift
public func schedule<StateType>(_ state: StateType, action: @escaping (StateType) -> Disposable) -> Disposable {
    if DispatchQueue.isMain {
        return action(state)
    }

    let cancel = SingleAssignmentDisposable()

    self._mainQueue.async {
        if cancel.isDisposed {
            return
        }

        cancel.setDisposable(action(state))
    }

    return cancel
}
```
So here we can see one conceptual difference: `MainScheduler` does not save an action's disposable to the `SingleAssignmentDisposable`. Was such a decision based on some assumptions or is it just a bug?

**RxSwift version**

 5.0.1

**Platform/Environment**

  - [x] iOS
  - [ ] macOS
  - [ ] tvOS
  - [ ] watchOS
  - [ ] playgrounds

**How easy is to reproduce? (chances of successful reproduce after running the self contained code)**

  - [x] easy, 100% repro
  - [ ] sometimes, 10%-100%
  - [ ] hard, 2% - 10%
  - [ ] extremely hard, %0 - 2%

**Xcode version**:

 11.3.1 (11C504)

**Installation method**:
  - [x] CocoaPods
  - [ ] Carthage
  - [ ] Git submodules

**I have multiple versions of Xcode installed**:
  (so we can know if this is a potential cause of your issue)
  - [ ] yes (which ones)
  - [x] no

**Level of RxSwift knowledge**:
  (this is so we can understand your level of knowledge
    and formulate the response in an appropriate manner)
  - [ ] just starting
  - [x] I have a small code base
  - [ ] I have a significant code base

**Comment by freak4pc** [comment]
Hey @valeriashershen, this is an interesting find :) Thanks for the thorough investigation. 

One thing to note here is that it's not entirely correct to say work is not disposed. It seems that specifically, if you schedule work on the Main Scheduler from a different execution context (in this case `DispatchQueue.global()`). 

I have to be honest in that I don't know this code well enough to say if it's a bug or not, but seeing this is the only difference between it and `MainScheduler.asyncInstance` (e.g. `SerialDispatchQueueScheduler(serialQueue: DispatchQueue.main)`), my guess would be this is some sort of an optimization.

Regardless, pinging @kzaher to take a look here if he has the time.

**Comment by freak4pc** [comment]
@kzaher trying the ping again :) 
Basically my main concern making this change would be that it _could_ cause regressions for people who accidentally count on this behavior.

**Comment by freak4pc** [comment]
@valeriashershen After thoroughly reviewing the code, I believe you're right and this seems like some sort of a bug. To not risk regressions to consumers of RxSwift 5, this fix will be part of RxSwift 6 (RC1 coming out next week), and hopefully enough consumers will be able to test it and make sure everything looks solid, given how popular `MainScheduler` is.

Fixed in 6a590c2b.
